### PR TITLE
Multi-resource timeline

### PR DESCRIPTION
**Pull request by eric-gunther**
image (8)

### CODE REVIEW

File: src/fhir/medications.ts [hunk 1]
```diff
@@ -124,6 +133,87 @@ export async function getBuilderMedications(
...
+
+    return resources.map(
+      (r) => new MedicationRequestModel(r, getIncludedResources(bundle))
+    );
+  } catch (e) {
+    throw errorResponse("Failed fetching medications for patient", e);
+  }
+}
+
+export async function getMedicationDispnseCommon(
```
Comment (WesleyKapow) [inline src/fhir/medications.ts:161]: typo, s/dispnse/dispense

File: src/components/content/timeline-2.0/helpers/filters.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  FilterItem,
+} from "@/components/core/filter-bar/filter-bar-types";
+import { SortOption } from "@/components/core/sort-button/sort-button";
+import { TimelineEventModel } from "@/fhir/models/timeline-event";
+import { compact, uniq } from "@/utils/nodash/fp";
+
+export const defaultTimelineSort: SortOption<TimelineEventModel> = {
+  display: "Date (New to Old)",
+  sorts: [{ key: "date", dir: "desc", isDate: true }],
+};
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/helpers/filters.ts:13]: Sort goes in `sorts.ts`

File: src/components/content/timeline-2.0/patient-timeline-columns.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    },
+    {
+      title: "Actor",
+      render: (timelineEvent) => (
+        <>
+          {timelineEvent.actor.map((row) => (
+            <div className="ctw-capitalize" key={row}>
+              {row.toLocaleLowerCase()}
+            </div>
+          ))}
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline-columns.tsx:31]: Use of `row` is confusing here as we're dealing with tables. Maybe `detail` instead? (this aligns with a suggestion I had to rename `actor` to `actorDetails`).

File: src/components/content/timeline-2.0/patient-timeline-columns.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        </>
+      ),
+    },
+    {
+      title: "Modifiers",
+      render: (timelineEvent) => (
+        <div>
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline-columns.tsx:38]: extra div not needed

File: src/components/content/timeline-2.0/patient-timeline.tsx [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+          defaultState: defaultConditionFilters,
+          filters: timelineFilters(timelineEventsQuery.data),
+        }}
+        sortOptions={{
+          defaultSort: defaultTimelineSort,
+          options: timelineSortOptions,
+          onChange: setSort,
+        }}
+      />
+      <Table
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline.tsx:60]: Use `ResourceTable` instead. See conditions and observations for examples of what this should look like.

File: src/components/content/timeline-2.0/patient-timeline.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  });
+  const openEncounterDetails = usePatientEncounterDetailsDrawer();
+  const openDiagnosticReportDetails = useObservationsDetailsDrawer();
+  const openMedicationDispenseDetails = useMedicationStatementDetailsDrawer();
+
+  return (
+    <div
+      className={cx(
+        className,
+        "ctw-scrollable-pass-through-height ctw-overflow-hidden"
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline.tsx:45]: `ctw-overflow-hidden` is redundant with what `ctw-scrollable-pass-through-height` provides.

File: src/components/content/timeline-2.0/patient-timeline.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+      )}
+    >
+      <ResourceTableActions
+        filterOptions={{
+          onChange: setFilters,
+          defaultState: defaultConditionFilters,
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline.tsx:51]: Should be default timeline filters.

File: src/components/content/timeline-2.0/patient-timeline.tsx [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+          defaultSort: defaultTimelineSort,
+          options: timelineSortOptions,
+          onChange: setSort,
+        }}
+      />
+      <Table
+        showTableHead={false}
+        isLoading={timelineEventsQuery.isLoading}
+        records={data}
+        columns={patientTimelineColumns(true)}
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline.tsx:64]: you won't want to hardcode view fhir button to true

File: src/components/content/timeline-2.0/patient-timeline.tsx [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        handleRowClick={(record) => {
+          if (record.model.constructor === EncounterModel) {
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline.tsx:66]: Use a switch statement on `record.model.resourceType`

File: src/components/content/timeline-2.0/patient-timeline.tsx [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+  medicationEventModel: MedicationDispenseModel | MedicationRequestModel;
+} & Pick<DrawerProps, "isOpen" | "onClose" | "onOpen" | "onAfterOpen">;
+
+const MedicationDrawerComponent = (props: MedicationDrawerComponentProps) => {
+  const { medicationEventModel } = props;
+  const medStatement = useQueryMedicationStatement(medicationEventModel.rxNorm);
+  if (medStatement?.data?.length) {
+    return <MedicationDrawer medication={medStatement.data[0]} {...props} />;
+  }
+  return <></>;
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/patient-timeline.tsx]: Pull out into a helper file.

File: src/components/content/timeline/patient-timeline.tsx [hunk 11]
```diff
@@ -59,6 +53,15 @@ const encounterData = (encounter: EncounterModel) => [
   { label: "Location", value: encounter.location },
   { label: "Participants", value: encounter.participants },
   { label: "Reason", value: encounter.reason },
-  { label: "Diagnosis", value: encounter.diagnosis },
+  { label: "Diagnosis", value: encounter.diagnoses?.join(", ") },
   { label: "Discharge Disposition", value: encounter.dischargeDisposition },
 ];
+
+export function usePatientEncounterDetailsDrawer() {
```
Comment (WesleyKapow) [inline src/components/content/timeline/patient-timeline.tsx:60]: Move to new `timeline-2.0/helpers/modal-hooks.ts` (to align with conditions)

File: src/hooks/use-medications.ts
Comment (WesleyKapow) [inline src/hooks/use-medications.ts:1]: Remove or cleanup copypasta comments :P

File: src/fhir/medications.ts
Comment (WesleyKapow) [inline src/fhir/medications.ts:1]: Functions that fetch a list of resources should be plural, so `getMedicationRequestsCommon` or maybe `getCommonMedicationRequests`, etc.

File: src/fhir/medications.ts [hunk 12]
```diff
@@ -124,6 +133,87 @@ export async function getBuilderMedications(
...
+        patientUPID: patient.UPID,
+        ...omitClientFilters(searchFilters),
+      }
+    );
+
+    return resources.map(
+      (r) => new MedicationRequestModel(r, getIncludedResources(bundle))
+    );
+  } catch (e) {
+    throw errorResponse("Failed fetching medications for patient", e);
```
Comment (WesleyKapow) [inline src/fhir/medications.ts:157]: Be more specific, "fetching medication requests" -- same goes for other copypasta

**Review by WesleyKapow — CHANGES_REQUESTED** [review]

File: src/components/content/conditions/helpers/sorts.ts [hunk 13]
```diff
@@ -37,3 +38,16 @@ export const conditionSortOptions: SortOption<ConditionModel>[] = [
...
+  sorts: [{ key: "date", dir: "desc", isDate: true }],
+};
+
+export const timelineSortOptions: SortOption<TimelineEventModel>[] = [
+  defaultTimelineSort,
+  {
+    display: "Date (Old to New)",
+    sorts: [{ key: "date", dir: "asc", isDate: true }],
+  },
+];
```
Comment (WesleyKapow) [inline src/components/content/conditions/helpers/sorts.ts:53]: This file is for conditions, move these to `timeline-2.0/helpers/sorts.ts`

File: src/fhir/medications.ts [hunk 14]
```diff
@@ -190,9 +190,9 @@ export async function getMedicationDispnseCommon(
 export async function getMedicationStatement(
```
Comment (WesleyKapow) [inline src/fhir/medications.ts:190]: `getMedicationStatements`

File: src/fhir/timeline-event.ts [hunk 15]
```diff
@@ -27,46 +27,40 @@ export function useTimelineEvents() {
...
+      applySorts(models, [
         { dir: "desc", key: "date", isDate: true },
-        { dir: "desc", key: "type", isDate: true },
+        { dir: "desc", key: "type" },
       ])
     );
+    // Disabling because including queries will cause infinite loop
+    // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [
     patientEncountersQuery.data,
```
Comment (WesleyKapow) [inline src/fhir/timeline-event.ts:55]: How about `queries.map(q => q.data)`?

**Review by WesleyKapow — CHANGES_REQUESTED** [review]

File: src/App.tsx [hunk 16]
```diff
@@ -58,54 +60,55 @@ type DemoComponent = {
...
+  //   render: () => <PatientConditions />,
+  //   title: "Patient Conditions",
+  // },
+  // {
+  //   render: () => <PatientDocuments />,
+  //   title: "Patient Documents",
+  // },
+  // { render: () => <PatientAllergies />, title: "Patient Allergies" },
+  // { render: () => <PatientImmunizations />, title: "Patient Immunizations" },
+  // { render: () => <PatientSearch />, title: "Patient Search" },
```
Comment (luthfib) [inline src/App.tsx:110]: Can you uncomment?

File: src/components/content/timeline-2.0/helpers/medication-detail.tsx [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+} & Pick<DrawerProps, "isOpen" | "onClose" | "onOpen" | "onAfterOpen">;
+
+const MedicationDrawerComponent = (props: MedicationDrawerComponentProps) => {
+  const { medicationEventModel } = props;
+  const medStatement = useQueryMedicationStatement(medicationEventModel.rxNorm);
+  if (medStatement.data?.length) {
+    return <MedicationDrawer medication={medStatement.data[0]} {...props} />;
+  }
+  return <></>;
+};
```
Comment (luthfib) [inline src/components/content/timeline-2.0/helpers/medication-detail.tsx:36]: Could probaly move this to src/components/content/timeline-2.0/helpers/modal-hooks.tsx

**Review by luthfib — DISMISSED** [review]
LGTM

**Review by luthfib — APPROVED** [review]

File: src/components/content/timeline-2.0/helpers/modal-hooks.tsx [hunk 18]
```diff
@@ -29,3 +35,33 @@ export const encounterData = (encounter: EncounterModel) => [
...
+  medicationEventModel: MedicationDispenseModel | MedicationRequestModel;
+} & Pick<DrawerProps, "isOpen" | "onClose" | "onOpen" | "onAfterOpen">;
+
+const MedicationDrawerComponent = (props: MedicationDrawerComponentProps) => {
+  const { medicationEventModel } = props;
+  const medStatement = useQueryMedicationStatement(medicationEventModel.rxNorm);
+  if (medStatement.data?.length) {
+    return <MedicationDrawer medication={medStatement.data[0]} {...props} />;
+  }
+  return <></>;
```
Comment (WesleyKapow) [inline src/components/content/timeline-2.0/helpers/modal-hooks.tsx:66]: You can `return null` in these cases.

**Review by WesleyKapow — APPROVED** [review]

**Review by luthfib — APPROVED** [review]

**Review by mrosata — APPROVED** [review]

**Pull request merged**